lib/metasploit/framework/credential_collection.rb

```diff
@@ -104,37 +104,34 @@ def prepend_cred(cred)
     self
   end
 
-  # Combines all the provided credential sources into a stream of {Credential}
-  # objects, yielding them one at a time
-  #
-  # @yieldparam credential [Metasploit::Framework::Credential]
-  # @return [void]
-  def each
+  def credentials
+    return @credentials if @credentials
+    @credentials = []
+    @credentials |= prepended_creds
+
     if pass_file.present?
       pass_fd = File.open(pass_file, 'r:binary')
     end
 
-    prepended_creds.each { |c| yield c }
-
     if username.present?
       if password.present?
-        yield Metasploit::Framework::Credential.new(public: username, private: password, realm: realm, private_type: private_type(password))
```

Comment:
This PR is marked as delayed.  TBH, beyond some minimal portion of the unit tests that I added, I may scrap the rest.  

The memory concerns are valid.  My main goal was to fix what I've described in #4541 which will likely only continue to get worse as more authbrute modules are add and as features are added/fixed to authbrute.